junit-jupiter-api/src/main/java/org/junit/jupiter/api/DynamicNode.java

```diff
@@ -40,6 +45,29 @@ public String getDisplayName() {
 		return this.displayName;
 	}
 
+	/**
+	 * Get the optional test source of this {@code DynamicNode}.
+	 *
+	 * @see #withTestSource(TestSource)
+	 */
+	public Optional<TestSource> getTestSource() {
+		return Optional.ofNullable(testSource);
+	}
+
+	DynamicNode setTestSource(TestSource testSource) {
+		this.testSource = testSource;
+		return this;
+	}
+
+	/**
+	 * Create new instance of this dynamic container or test with a custom test source.
+	 *
+	 * @param testSource custom test source instance to associate with this node
+	 * @return new node instance
+	 * @see #getTestSource()
+	 */
+	public abstract DynamicNode withTestSource(TestSource testSource);
```

Comment:
How about just implementing this instead of making it `abstract`?

Or... do you foresee introducing additional methods in a pseudo-builder pattern style in the future?